### PR TITLE
[SPARK-34033][R] SparkR Daemon Initialization

### PR DESCRIPTION
**Pull request by WamBamBoozle**
### What changes were proposed in this pull request?

A way of letting sparkR users preload libraries and other initialization
in the daemon before it starts forking. A full description can be found
in [docs/sparkr.md#daemon-initialization](/WamBamBoozle/spark/blob/daemon_init/docs/sparkr.md#daemon-initialization).

### Why are the changes needed?

Performance

### Does this PR introduce _any_ user-facing change?

Yes. It introduces two new sparkR session parameters,
spark.r.daemonInit and spark.r.daemonTimeout, as described in
[docs/configuration.md#sparkr](/WamBamBoozle/spark/blob/daemon_init/docs/configuration.md#sparkr).

### How was this patch tested?

added pkg/tests/fulltests/test_daemon_initialization.R

### CODE REVIEW

File: docs/configuration.md [hunk 1]
```diff
@@ -2824,6 +2824,28 @@ External users can query the static sql config values via `SparkSession.conf` or
...
+</tr>
+<tr>
+  <td><code>spark.r.daemonTimeout</code></td>
+  <td>10000</td>
+  <td>
+    What is the upper limit in milliseconds that we allow for daemon
+    initialization? In other words, how long should the spark executor
+    wait until it decides that it can't launch the SparkR daemon?
+  </td>
+  <td>3.2.0</td>
```
Comment (WamBamBoozle) [inline docs/configuration.md:2847]: Is this the correct release to put here?

**Comment by WamBamBoozle** [comment]
Greetings @HyukjinKwon

Thanks so much for your review of #31072! I had to close that PR as I forgot rebasing git causes bad things to happen in github.

Regarding parameterizing the path to the core components of sparkR, namely daemon.R and worker.R. I think this is an excellent solution if your objective is to implement coverage analysis. I just want my users to routinely move their initialization into the daemon.

I'm looking at python/test_coverage/coverage_daemon.py

I agree that spark.python.daemon.module is more general and more powerful than what I am proposing. This PR is for fixing one thing, whereas what you did was a very general solution to a different problem. I don't want to warn my users that this is for advanced users only. The concept is simple: put your initialization code here.

I propose we provide

spark.r.daemon.finalize
spark.r.daemon.initialize
spark.r.worker.finalize
spark.r.worker.initialize
where the value of each of those are strings of R code

that way we've achieved the same ends you have without forcing the user to understand the minutia of how and what to load. If you agree, I will change this PR to be the implementation of spark.r.daemon.initialize (instead of what it is now, spark.r.daemonInit). I would then open another ticket to implement the other flags.

**Comment by HyukjinKwon** [comment]
I still think it's better to have one solution to cover many cases. Having `spark.r.daemonInit` will fix one arguably corner case. You can wait for other committers to comment if you'd like to stick with the current approach.

**Comment by WamBamBoozle** [comment]
I think the solution you suggest would intimidate my users so yes, I would love a second opinion.

File: R/pkg/inst/worker/daemon.R [hunk 2]
```diff
@@ -32,6 +32,9 @@ inputCon <- socketConnection(
 
 SparkR:::doServerAuth(inputCon, Sys.getenv("SPARKR_WORKER_SECRET"))
 
+# Application-specific daemon initialization. Typical use is loading libraries.
+eval(parse(text = Sys.getenv("SPARKR_DAEMON_INIT")))
+
```
Comment (WamBamBoozle) [inline R/pkg/inst/worker/daemon.R:37]: To give you a sense of the cost of this, consider
```
> microbenchmark(NULL, times = 999999)
Unit: nanoseconds
 expr min lq     mean median uq   max  neval
 NULL   2  4 4.328575      4  4 13885 999999
> microbenchmark(eval(parse(text = "NULL")), times=9999)
Unit: microseconds
                       expr    min      lq    mean median     uq      max neval
 eval(parse(text = "NULL")) 31.489 34.0545 40.1457 35.938 38.496 9370.646  9999
> 
```
So, on my 2017 MacBook pro, it takes 34 microseconds.

For comparison, consider the 6 milliseconds we use up loading worker.R at every fork.

File: R/run-tests.sh [hunk 3]
```diff
@@ -30,11 +30,22 @@ if [[ $(echo $SPARK_AVRO_JAR_PATH | wc -l) -eq 1 ]]; then
...
+NOT_CRAN=true \
+SPARK_TESTING=1 \
+$FWDIR/../bin/spark-submit \
+--conf spark.driver.extraJavaOptions="-Dio.netty.tryReflectionSetAccessible=true" \
+--conf spark.executor.extraJavaOptions="-Dio.netty.tryReflectionSetAccessible=true" \
+--conf spark.hadoop.fs.defaultFS="file:///" \
+--conf spark.r.daemonInit='testInit <- "wow"' \
+--driver-java-options "-Dlog4j.configuration=file:$FWDIR/log4j.properties" \
+$JARS \
+$FWDIR/pkg/tests/run-all.R 2>&1 | tee -a $LOGFILE
```
Comment (WamBamBoozle) [inline R/run-tests.sh:47]: The only thing new here is "spark.r.daemonInit" -- the rest is just a refactoring of a duplication above.

**Comment by srowen** [comment]
I dont' know much about R - why does this help improve performance?

**Comment by WamBamBoozle** [comment]
> I dont' know much about R - why does this help improve performance?

It is like moving the invariant expression out of the loop. It is the difference between loading supporting libraries once vs loading them millions of times.

It is described in detail in [docs/sparkr.md](https://github.com/WamBamBoozle/spark/blob/daemon_init/docs/sparkr.md#daemon-initialization)

**Comment by srowen** [comment]
I see. This sounds a bit like proposals for executor-wide initialization, though this is a bit more acute as a new R process spawns for every task. I wonder if there are side effects, like, one-time init from one job being picked up in later ones and that causing surprise. I haven't thought that through though.

**Pull request closed without merging**